File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int[] inorder;
    int[] preorder;
    Map<Integer, Integer> map;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.preorder = preorder;
        this.inorder = inorder;
        int n = preorder.length;
        map = new HashMap<>();
        for(int i = 0; i < n; i++) map.put(inorder[i], i);
        return dfs(0, n - 1, 0, n - 1);
    }
    public TreeNode dfs(int pstart, int pend, int istart, int iend) {
        if(pstart > pend) return null;
        TreeNode root = new TreeNode(preorder[pstart]);
        int rootIndex = map.get(preorder[pstart]);
        int leftCount = rootIndex - istart;
        root.left = dfs(pstart + 1, pstart + leftCount, istart, rootIndex - 1);
        root.right = dfs(pstart + leftCount + 1, pend, rootIndex + 1, iend);
        return root;
    }
}

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        map = {}
        for i, v in enumerate(inorder):
            map[v] = i
        n = len(preorder)
        print(n)
        def dfs(pstart, pend, istart, iend):
            if pstart > pend:
                return None
            root = TreeNode(preorder[pstart])
            root_index = map[preorder[pstart]]
            left_cnt = root_index - istart
            root.left = dfs(pstart + 1, pstart + left_cnt, istart, root_index - 1)
            root.right = dfs(pstart + left_cnt + 1, pend, root_index + 1, iend)
            return root
        return dfs(0, n - 1, 0, n - 1)
    